fc/code_generation.py

```diff
@@ -195,45 +192,47 @@ def derivative_name(deriv):
             'var_name': symbol_name(state),
             'deriv_name': derivative_name(state),
             'initial_value': state.initial_value,
-            'var_names': model.get_ontology_terms_by_symbol(state, oxmeta),
+            'var_names': model.get_ontology_terms_by_symbol(state, OXMETA_NS),
         })
 
     # Create parameter information dicts, and map of parameter symbols to their indices
     parameter_info = []
     parameter_symbols = {}
     for i, parameter in enumerate(parameters):
-        symbol = model.get_symbol_by_ontology_term(*parameter)
+        symbol = model.get_symbol_by_ontology_term(parameter.rdf_term)
         parameter_info.append({
             'index': i,
-            'annotation': parameter,
+            'local_name': parameter.local_name,
             'var_name': symbol_name(symbol),
             'initial_value': model.get_value(symbol),
         })
         parameter_symbols[symbol] = i
 
     # Create output information dicts
-    # Each output is associated either with a symbol or a parameter.
+    # Each output is associated either with a symbol or a list thereof.
     output_info = []
+    output_symbols = set()
     for i, output in enumerate(outputs):
-
-        # Allow special output value 'state_variable'
-        # TODO Replace this with a more generic implementation
-        if output[1] == 'state_variable':
-            var_name = [{'index': x['index'], 'var_name': x['var_name']}
-                        for x in state_info]
-            parameter_index = None
-            length = len(state_info)
+        term = output.rdf_term
+        symbols = get_variables_transitively(model, term)
+        output_symbols.update(symbols)
+        if len(symbols) == 0:
+            raise ValueError('No variable annotated as {} found'.format(term))
+        elif len(symbols) == 1:
+            length = None  # Not a vector output
```

Comment:
This `length = None` distinguishes between scalars and length-1 vectors, but the call to `get_variables_transitively` does not?

Reply:
You find out whether it's a scalar or vector output by seeing how many variables are annotated with the term.

Reply:
As in, there is no concept of a length-1 vector (at present).